src/details/ArborX_DetailsDistributedTreeImpl.hpp

```diff
@@ -671,6 +678,141 @@ DistributedTreeImpl<DeviceType>::queryDispatch(
+  // - Send the result back to the process owning the predicates.
+
+  // Find the ranks and indices for the nearest queries using the overload not
+  // taking a callback.
+  queryDispatchImpl(NearestPredicateTag{}, tree, space, queries, indices,
+                    offset, ranks);
+
+  Kokkos::Profiling::pushRegion(
+      "ArborX::DistributedTree::query::nearest::postprocess_callback");
+
+  // Send the predicate-primitive pairs to the process where the match was
+  // found.
+  auto comm = tree.getComm();
+  int comm_rank;
+  MPI_Comm_rank(comm, &comm_rank);
+
+  using Access = AccessTraits<Predicates, PredicatesTag>;
+  using Query = typename AccessTraitsHelper<Access>::type;
+
+  Kokkos::View<QueriesWithIndices<Query> *, typename DeviceType::memory_space>
+      exported_queries_with_indices(
+          Kokkos::view_alloc(
+              space, Kokkos::WithoutInitializing,
+              "ArborX::DistributedTree::query::exported_queries_with_indices"),
+          ranks.size());
+  Kokkos::parallel_for(
+      "ArborX::DistributedTree::query::zip_queries_and_primitives",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, Access::size(queries)),
+      KOKKOS_LAMBDA(int i) {
+        using index_type = typename OffsetView::value_type;
+        for (index_type i = offset(q); i < offset(q + 1); ++i)
+          exported_queries_with_indices(i) = {Access::get(queries, q), q,
+                                              indices(i)};
+      });
+
+  Distributor<DeviceType> distributor(comm);
+  auto const n_imports = distributor.createFromSends(space, ranks);
+
+  Kokkos::View<QueriesWithIndices<Query> *, typename DeviceType::memory_space>
+      imported_queries_with_indices(
+          Kokkos::view_alloc(
+              space, Kokkos::WithoutInitializing,
+              "ArborX::DistributedTree::query::imported_queries_with_indices"),
+          n_imports);
+
+  sendAcrossNetwork(space, distributor, exported_queries_with_indices,
+                    imported_queries_with_indices);
+
+  // Execute the callback on the process owning the primitives.
+  OutputView remote_out(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing,
+                         "ArborX::DistributedTree::query::remote_out"),
+      n_imports);
+  KokkosExt::reallocWithoutInitializing(space, indices, n_imports);
+  Kokkos::parallel_for(
+      "ArborX::DistributedTree::query::execute_callbacks",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0,
+                                          imported_queries_with_indices.size()),
+      KOKKOS_LAMBDA(int i) {
+        callback(imported_queries_with_indices(i).query,
+                 imported_queries_with_indices(i).primitive_index,
+                 [&](typename OutputView::value_type const &value) {
+                   remote_out(i) = value;
+                   indices(i) = imported_queries_with_indices(i).query_id;
+                 });
+      });
```

Comment:
This seems wrong for a few reasons.

First, it assumes that callback will output a single element for each match. In practice, we don't know how many elements the callback will spit out per query. It could be 0, or it could be more than 1. So we need to do a two pass here, similar to the `CrsGraphWrapperImpl`. I'm not sure what is the easiest way to do it without massive code duplication.

Second, it assumes inline callback and not a post one. We could try to support post callback, but its meaning would have to be changed as we cannot execute it on all of the results coming from different ranks.

Reply:
We could also just assume that the callback only returns a single element to get started, right?

Reply:
We could. The thing is, this condition is undetectable, so the failure for a user would likely be an undefined behavior or a segfault. Which in the distributed setting would be tricky to debug.

Reply:
I added an assertion.